lib/eslint/eslint.js

```diff
@@ -0,0 +1,471 @@
+ * @property {LintMessage[]} messages All of the messages for the result.
+ * @property {number} errorCount Number of errors for the result.
+ * @property {number} warningCount Number of warnings for the result.
+ * @property {number} fixableErrorCount Number of fixable errors for the result.
+ * @property {number} fixableWarningCount Number of fixable warnings for the result.
+ * @property {string} [source] The source code of the file that was linted.
+ * @property {string} [output] The source code of the file that was linted, with as many fixes applied as possible.
+ * @property {DeprecatedRuleInfo[]} usedDeprecatedRules The list of used deprecated rules.
+ */
+
+/**
+ * A formatter object.
+ * @typedef {Object} Formatter
+ * @property {(results: LintResult[]) => string} format The main formatter method.
+ */
+
+/**
+ * Private members for the `ESLint` instance.
+ * @typedef {Object} ESLintPrivateMembers
+ * @property {CLIEngine} cliEngine The wrapped CLIEngine instance.
+ * @property {ESLintOptions} options The options used to instantiate the ESLint instance.
+ */
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * The map with which to store private class members.
+ * @type {WeakMap<ESLint, ESLintPrivateMembers>}
+ */
+const privateMembersMap = new WeakMap();
+
+/**
+ * Checks if a plugin object is valid.
+ * @param {PluginElement} plugin The plugin to check.
+ * @returns {boolean} Whether te plugin is valid or not.
+ */
+function isValidPluginObject(plugin) {
+    return typeof plugin === "object" && !Array.isArray(plugin) && plugin !== null && plugin.id && plugin.definition;
+}
+
+/**
+ * Normalizes an array of plugins to their respective IDs.
+ * @param {string[]|PluginElement[]} plugins An array of plugins to normalize.
+ * @returns {string[]} The normalized array of plugins.
+ */
+function normalizePluginIds(plugins) {
+    return plugins.map(p => (typeof p === "string" ? p : p.id));
+}
+
+/**
+ * Validates and normalizes options for the wrapped CLIEngine instance.
+ * @param {ESLintOptions} options The options to process.
+ * @returns {ESLintOptions} The normalized options.
+ */
+function processOptions({
+    allowInlineConfig = true,
+    baseConfig = null,
+    cache = false,
+    cacheLocation = ".eslintcache",
+    configFile = null,
+    cwd = process.cwd(),
+    envs = [],
+    extensions = null, // TODO: Should we update the CLIEngine check to account for empty arrays?
+    fix = false,
+    fixTypes = [], // TODO: We don't currently set a default value for this. Doing so changes the behavior.
+    globals = [],
+    globInputPaths = true,
+    ignore = true,
+    ignorePath = null,
+    ignorePattern = [],
+    parser = null, // Question: if this is set to a default value it ends up overriding the value set in the config. That seems like unexpected behavior to me.
```

Comment:
> Question: if this is set to a default value it ends up overriding the value set in the config. That seems like unexpected behavior to me.

All CLI options override the settings in config files. Therefore, if `options.parser` is present, it overrides the value of config files. On the other hand, `options.baseConfig.parser` doesn't override the value of config files.

Maybe we should aggregate CLI options that overrides config files to one such as `cliOptions` because now the class name is not `CLIEngine`.

Reply:
How about https://github.com/eslint/eslint/commit/c96d18dda71a176ee691f560fa7c8bdad329ff0c?

- It adds `overrideConfig` that is similar to `baseConfig`.
- It renames `configFile` to `overrideConfigFile` to be clear.
    1. The loaded config data from `overrideConfigFile` will be added to the tail of every config array.
    2. The normalized config data from `overrideConfig` will be added to the tail of every config array after `overrideConfigFile` is handled.
- It removes `allowInlineConfig`, `envs`, `globals`, `ignorePattern`, `parser`, `parserOptions`, and `rules` because of preferring the corresponded fields of `overrideConfig`.
- It changes `plugins` to only for `addPlugin`. It doesn't add the `plugins` setting of configurations as same as the current `addPlugin`. The `plugins` of CLIEngine moved to `overrideConfig.plugins`.
- The `ESLint` constructor shows error messages to show the corresponded fields of `overrideConfig` if those removed options are present.

Reply:
The `overrideConfigFile` may not be needed because we can use  `overrideConfig.extends` instead.